### PR TITLE
[TST] Add integration test

### PR DESCRIPTION
**Pull request by eurunuela**
Closes none

This PR adds an integration test to make sure we don't break anything with other PRs. It uses inputs and outputs from the 5 echo integration test in tedana.

### CODE REVIEW

**Comment by eurunuela** [comment]
@tsalo when you restructured the code to look like sklearn, did you make any changes that could affect the functionality? I think I'm doing the exact same as maPCA in tedana here, but the number of components I'm getting is different.

**Comment by tsalo** [comment]
I don't think any of my changes should have affected the results, _except_ for the implementation of `inverse_transform()` (I think I have a comment about that in the docstring).

Could it be the normalization? Normalization was disabled in tedana and is done in this repo's version.

**Comment by eurunuela** [comment]
I added the same normalization to both maPCA and tedana.

`inverse_transform()` is never used in maPCA, right? Or am I missing something?

**Comment by eurunuela** [comment]
Okay, I'll have a look at that. I went crazy yesterday trying to make them match but I ran out of ideas. I'll let you know once I check the reshaping.

**Comment by eurunuela** [comment]
The reshaping is not the problem. I don't know why we're getting different solutions now.

**Comment by notZaki** [comment]
If the output in `test/resources` was copied from tedana, then maybe it's because tedana uses `aic` instead of `mdl` to estimate the number of dimensions.

**Comment by notZaki** [comment]
Also, `X` should be normalized in `MovingAveragePCA.transform` if the goal is to agree with tedana:
https://github.com/ME-ICA/mapca/blob/58cbeea64bb3ed84c5bd0bf23807185f0125566b/mapca/mapca.py#L286

**Comment by eurunuela** [comment]
> If the output in `tests/resources` was copied from tedana, then maybe it's because tedana uses `aic` instead of `mdl` to estimate the number of components.

Tedana uses `mdl` by default, of that I'm sure.

> Also, I think `X` should be normalized in `MovingAveragePCA.transform` if the goal is to agree with tedana:
> https://github.com/ME-ICA/mapca/blob/58cbeea64bb3ed84c5bd0bf23807185f0125566b/mapca/mapca.py#L286

You're right, we're missing the normalization on that step, but that should not affect the estimation of the number of components.

**Comment by notZaki** [comment]
Another reason the values might not agree is that the PCA step mutates the data:
https://github.com/ME-ICA/mapca/blob/58cbeea64bb3ed84c5bd0bf23807185f0125566b/mapca/mapca.py#L214-L215

In tedana, [the mutated `data` (or `X`) is used](https://github.com/ME-ICA/tedana/blob/dae9f20d80c6aa365d443faac55c2db3b7b94374/tedana/decomposition/ma_pca.py#L607-L611) to estimate `u` (or `X_new`), whereas here the original `X` is being used.

**Comment by eurunuela** [comment]
Yes, I'm using the normalized `data` in this PR.  I've reverted back to the way tedana reshapes the data and now the results are closer, but still not identical. 

Apparently, the normalized `data` in [tedana lines 607-611](https://github.com/ME-ICA/tedana/blob/dae9f20d80c6aa365d443faac55c2db3b7b94374/tedana/decomposition/ma_pca.py#L607-L611) is not exactly the same as `X` in [maPCA line 286](https://github.com/ME-ICA/mapca/blob/58cbeea64bb3ed84c5bd0bf23807185f0125566b/mapca/mapca.py#L286) with 

```python
X = self.scaler_.fit_transform(X.T).T
X_new = np.dot(np.dot(X, self.components_.T), np.diag(1.0 / self.explained_variance_))
```

**Comment by notZaki** [comment]
Are `data` and `X` identical prior to tedana line 607? If they're different after that, then that's to be expected because `PCA.fit` will modify `data`.

Running the test locally after [this change](https://github.com/notZaki/mapca/commit/0b3983b14fdb06e25d51025fd75c8413b3470491) made the integration test pass. Though this obviously won't work as the final product because `MovingAveragePCA.transform` is not being used at all.

**Comment by eurunuela** [comment]
Thank you @notZaki ! I think we can implement that change for the moment so we can keep working on the other PRs. We can find a fix for `MovingAveragePCA.transform` later.

**Comment by eurunuela** [comment]
Let me know what you think @notZaki @tsalo 

I think once we merge this we will go faster with the remaining PRs and issues.

**Comment by notZaki** [comment]
Although the values might agree, I think the code here makes more sense than what tedana is doing, i.e. we should use the original `X` instead of the one modified after `PCA.fit`. But I guess the point is just to make sure we're not changing the results in future PRs, so it's fine either way.

Regarding this PR: Will the data stay in `tests/resources` or will it be uploaded elsewhere (OSF?) and downloaded on-demand during tests.

**Comment by eurunuela** [comment]
> Regarding this PR: Will the data stay in `tests/resources` or will it be uploaded elsewhere (OSF?) and downloaded on-demand during tests?

As you wish. I have no strong opinions about it. The folder is currently 25 MB, which isn't too big. I do think OSF is the "cleaner", more elegant way of dealing with the data though. 

How about we open an issue for that and deal with it on another PR? I just don't this PR to delay things anymore.

**Comment by notZaki** [comment]
The decision to include those large files should be made before merging the PR because once those files get merged, it's going to be difficult to completely remove them since git will always store them in its history---the total repo size can only increase over time. 

I think resolving this now will be better than having to tackle this in the future.
If you can make an OSF repo and upload the resource files, then it should be relatively quick to have a function that downloads those resources prior to testing.

**Comment by eurunuela** [comment]
I didn't know that about GitHub. I'll make an OSF repo tomorrow.

**Review by notZaki — APPROVED** [review]

**Comment by notZaki** [comment]
LGTM!

The resources are still in the git history, but I think if the PR is merged using the `Squash and merge` option then it should be fine.

**Comment by eurunuela** [comment]
Thank you for you help @notZaki !

@tsalo are you happy with the changes? I want to make sure we're all happy before merging.

**Comment by tsalo** [comment]
It looks like everything makes sense and I'm really happy it's working! I do have a couple of questions about changes to the core code.

Just to summarize the changes to the core code:
1. Transpose the inputs to the scaler so it scales along the appropriate dimension.
    - It looks like this fixes a bug that I must have introduced in my initial conversion.
2. Feed in flattened, but _unmasked_ data to the scikit-learn style transformer, rather than flattened and masked data.
    - Was the old way not working? I'm sure it's fine either way, although if we're not able to remove the downsampling then we'll probably want to just feed the full data and mask images into the sklearn-style Transformer eventually. Depending on what the `nilearn` devs want, of course.
3. Calculate `self.u` in `_fit()` instead of calculating it (not as an attribute) in `transform()`.
    - Is this just because it makes usage simpler? I know we want `u`, so I'm guessing the goal is just to use `fit()` without needing `transform()`? It moves us further away from the scikit-learn approach, but given the restricted use-case for this method that is probably fine.

**Review by tsalo — APPROVED** [review]

**Comment by eurunuela** [comment]
> 1. Transpose the inputs to the scaler so it scales along the appropriate dimension.
>    
>    * It looks like this fixes a bug that I must have introduced in my initial conversion.

This is actually related to the discussion in [tedana #636](https://github.com/ME-ICA/tedana/pull/636) about how we should be z-scoring in the temporal domain rather than the spatial one.

> 2. Feed in flattened, but _unmasked_ data to the scikit-learn style transformer, rather than flattened and masked data.
>    
>    * Was the old way not working? I'm sure it's fine either way, although if we're not able to remove the downsampling then we'll probably want to just feed the full data and mask images into the sklearn-style Transformer eventually. Depending on what the `nilearn` devs want, of course.

Apparently, running maPCA with the `apply_mask()` function in nilearn didn't yield results identical to the `tedana` outcome. That's why I reverted back to how `tedana` applies the mask in `mapca()`.

> 3. Calculate `self.u` in `_fit()` instead of calculating it (not as an attribute) in `transform()`.
>    
>    * Is this just because it makes usage simpler? I know we want `u`, so I'm guessing the goal is just to use `fit()` without needing `transform()`? It moves us further away from the scikit-learn approach, but given the restricted use-case for this method that is probably fine.

This is because the data is normalized during the PCA step and this normalized data is not accessible from `transform()`. We can think of a better way of doing it but right now that is the quickest fix I could think of.

Feel free to open issues related to these core changes. I will merge this PR now so we can keep working on the remaining open PRs. Thanks again @tsalo and @notZaki !
